### ISSUE TITLE
First-class procedures with unevaluated inputs

### ISSUE THREAD

**Issue by rdococ** [jmoenig/Snap#2269]
Input types don't really matter with first-class procedures - not that they really did anyway because they merely serve as guides, but the "unevaluated" input type has different behavior - namely, wrapping the input in a thunk that you can apply to evaluate whenever you want. However, first-class procedures don't have this functionality.

![untitled script pic](https://user-images.githubusercontent.com/3169568/49084598-d54d1a80-f247-11e8-8a78-336dc0ea2680.png)

Some of the use cases for this could be to invoke methods of other sprites that use unevaluated inputs. Plus, I think that everything behavioral that a custom block can do, a first-class procedure should be able to do as well.

You should also be able to do this with explicit procedure inputs by right-clicking and checking"Unevaluated".

**Comment by brianharvey** [comment]
I agree that it should be possible to type an explicit formal parameter of an anonymous procedure, at least for the procedure/unevaluated types.

I don't think we should hair up implicit empty-slot inputs, though.  Their whole purpose is to allow beginners to use higher order functions _without knowing about_ anonymous procedures.

By the way, I changed the title of this issue.  "First class" is a property of a data type, not of a particular member of that data type.  So _procedures_ are first class, because a procedure can be a value of a variable, an input to a procedure, ..., and exist on its own without a name.  The specific procedures that don't have names are called "anonymous."  (Same with any data type; you just don't hear people talk much about "anonymous integers" because even Fortran had those, so the fact that an integer can be anonymous seems "obvious" to Fortran/C/Java/Python/etc. programmers even though there's nothing obvious about it.  (For example, in Scratch 1.4 you couldn't put a Boolean value in a variable -- it was changed to a number.)
